File: lab4/Vector2d.java
package agh.cs.lab4;

public class Vector2d {
    public final int x;
    public final int y;

    public Vector2d(int x, int y){
        this.x = x;
        this.y = y;
    }

    public boolean equals(Object other){
        if(this == other)
            return true;

        if(!(other instanceof Vector2d))
            return false;

        Vector2d that = (Vector2d) other;

        return this.x == that.x && this.y == that.y;
    }
}

File: lab2/src/main/kotlin/agh/ics/oop/Vector2d.kt
package agh.ics.oop

import kotlin.math.max
import kotlin.math.min

class Vector2d(val x: Int,val y: Int){
    override fun toString(): String = "($x, $y)"
    fun precedes(other: Vector2d) = (this.x <= other.x) && (this.y <= other.y)
    fun follows(other: Vector2d) = (this.x >= other.x) && (this.y >= other.y)
    fun upperRight(other: Vector2d) = Vector2d(max(this.x, other.x), max(this.y, other.y))
    fun lowerLeft(other: Vector2d) = Vector2d(min(this.x, other.x), min(this.y, other.y))
    fun add(other: Vector2d) = Vector2d(this.x + other.x, this.y + other.y)
    fun subtract(other: Vector2d) = Vector2d(this.x - other.x, this.y - other.y)
    override fun equals(other: Any?): Boolean {
        if (other is Vector2d){
            return (this.x == other.x) && (this.y == other.y)
        }
        return false
    }
    fun opposite() = Vector2d(-this.x, -this.y)
}